engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala

```diff
@@ -16,31 +15,42 @@ import pl.touk.nussknacker.engine.flink.api.process.{FlinkCustomNodeContext, Fli
 import pl.touk.nussknacker.engine.flink.util.keyed.{KeyedValue, KeyedValueMapper}
 import pl.touk.nussknacker.engine.kafka.{KafkaConfig, PartitionByKeyFlinkKafkaProducer, PreparedKafkaTopic}
 
-class KafkaAvroSink(preparedTopic: PreparedKafkaTopic, versionOption: SchemaVersionOption, key: LazyParameter[AnyRef], value: LazyParameter[AnyRef],
-                    kafkaConfig: KafkaConfig, serializationSchemaFactory: KafkaAvroSerializationSchemaFactory,
-                    schema: NkSerializableAvroSchema, runtimeSchema: Option[NkSerializableAvroSchema],
-                    clientId: String, validationMode: ValidationMode)
+class KafkaAvroSink(preparedTopic: PreparedKafkaTopic,
+                    versionOption: SchemaVersionOption,
+                    key: LazyParameter[AnyRef],
+                    sinkValue: AvroSinkValue,
+                    kafkaConfig: KafkaConfig,
+                    serializationSchemaFactory: KafkaAvroSerializationSchemaFactory,
+                    schema: NkSerializableAvroSchema,
+                    runtimeSchema: Option[NkSerializableAvroSchema],
+                    clientId: String,
+                    validationMode: ValidationMode)
   extends FlinkSink with Serializable with LazyLogging {
 
   import org.apache.flink.streaming.api.scala._
 
   // We don't want serialize it because of flink serialization..
   @transient final protected lazy val avroEncoder = BestEffortAvroEncoder(validationMode)
 
-  override def registerSink(dataStream: DataStream[InterpretationResult], flinkNodeContext: FlinkCustomNodeContext): DataStreamSink[_] = {
-    dataStream
-      .map(_.finalContext)
-      .map(new KeyedValueMapper(flinkNodeContext.lazyParameterHelper, key, value))
+  override def registerSink(dataStream: DataStream[InterpretationResult], flinkNodeContext: FlinkCustomNodeContext): DataStreamSink[_] =
+    toValueWithContext(dataStream.map(_.finalContext), flinkNodeContext)
       .map(new EncodeAvroRecordFunction(flinkNodeContext))
       .filter(_.value != null)
       .addSink(toFlinkFunction)
-  }
 
   /**
    * Right now we support it incorrectly, because we don't use default sink behavior with expression..
    */
   override def testDataOutput: Option[Any => String] = Some(value => Option(value).map(_.toString).getOrElse(""))
 
+  private def toValueWithContext(ds: DataStream[Context], flinkNodeContext: FlinkCustomNodeContext): DataStream[ValueWithContext[KeyedValue[AnyRef, AnyRef]]] =
+    sinkValue match {
+      case AvroSinkPrimitiveValue(value) =>
+        ds.map(new KeyedValueMapper(flinkNodeContext.lazyParameterHelper, key, value))
```

Comment:
Let's add TODO to KeyedValueMapper to change it also to flatMap and handle errors properly?

Reply:
:+1: 
https://github.com/TouK/nussknacker/pull/1405/commits/9f8cf216e127ebb7af131f51b2e1a2bec6283526